app/models/user.rb

```diff
@@ -1,6 +1,6 @@
 class User < ApplicationRecord
   has_and_belongs_to_many :lists
-  has_many :tasks
+  has_many :tasks, dependent: :restrict_with_error
```

Comment:
I think I can finally articulate our differing opinions here!

My preference is to use `dependent: :destroy` because I like the idea of having a graph that can clean itself up. Your preference is to use `dependant: :restrict_with_error` because you like the idea of data as being paramount. We both think graphs should clean themselves up, and we both think data is paramount, but our priorities are different.

How do we resolve this?

Reply:
I wavered on this too. I had it set to `dependent: :destroy` while I was tinkering with dummy data, but then considered the safety of exposing user destruction:
1. Use `dependent: :destroy`, but then ensure that every use of `user.destroy` is okay.
1. Use `dependent: :restrict_with_error` and then code a sanctioned bypass if needed, `user.really_destroy_user_and_all_tasks`

Reply:
Indeed.

A couple of reasons why I think `dependent: :destroy` is superior:

1. If I want to destroy the user I have to go destroy the tasks first. Lets say the graph was bigger and I needed to delete multiple descendants. Every time I manually go `foo.destroy` I risk deleting the wrong foo. If I only have to go `user.destroy` then I can have confidence in the graph cleaning up itself

2. Backups. We have them.

3. Only developers with console access can do this. It can't be done by a user through the API so the barrier to deleting is already pretty high.

4. Foreign key constraints help us. I need to double check, but I'm pretty sure we haven't specified cascading rules in our migrations and it defaults to restrict. So we actually have this behaviour at the database level. The app level just makes it convenient.

Are you convinced? 😉. I'd be happy to chat offline.

Reply:
Actioned in ffbd61c.